lib/split/trial.rb

```diff
@@ -3,15 +3,30 @@ module Split
   class Trial
     attr_accessor :experiment
     attr_accessor :metadata
+    attr_accessor :version
 
     def initialize(attrs = {})
       self.experiment   = attrs.delete(:experiment)
       self.alternative  = attrs.delete(:alternative)
       self.metadata  = attrs.delete(:metadata)
+      self.version = attrs.delete(:version)
 
       @user             = attrs.delete(:user)
       @options          = attrs
 
+      if self.version.nil?
+        if @user
+          key_for_experiment = @user.keys.find { |k| k.match(Regexp.new("^#{experiment.name}"))}
+          if key_for_experiment
+            self.version = key_for_experiment.split(":").last
```

Comment:
Is the sort order of the keys guaranteed to be stable here?  If not we could end up returning an older version by mistake.  I *think* this is going to be defined by what Redis returns. It may be better to sort them explicitly.  I'm assuming that we don't expect to have a lot of versions.

Reply:
a user is only ever in a single version of an experiment at a time. So for the purposes of getting the first key for an experiment we should be fine.

Reply:
Sorry, this comment is ambiguous because I left it on the wrong line.  I was referring to:
```ruby
key_for_experiment = @user.keys.find { |k| k.match(Regexp.new("^#{experiment.name}"))}
```
If I remember the Split API correctly, the `@user` attribute is actually a Redis connection object.  So `@user.keys` will return every key in the redis instance. The returned value would be the first key that matches the experiment name, so this line functioning correctly would require that Redis's ordering of keys did so resulting in the latest experiment being in that list before all others.  Is that stable? As in can we rely on it?  For instance, if we failed over from one redis server to another, is the order of keys guaranteed to be stable between servers?

Reply:
Or is the `@user` attribute something else? I know there's  a place where it's a Redis connection object which makes this confusing.

I hate their naming conventions :-/

Reply:
I figured it out. The `Split::User` class has the following setup:

```ruby
extend Forwardable
def_delegators :@user, :keys, :[], :[]=, :delete
```

So the `key` method will be delegated to the `@user` attribute.  This attribute is set to the persistence adapter. So not the Redis connection object directly, but an instance of the `Split::Persistence::RedisAdapter` class.  The `keys` method of this class doesn't quite do what I said earlier:
```ruby
  def keys
    Split.redis.hkeys(redis_key)
  end
```
So it isn't returning all known keys, but rather grabbing a single key, which is expected to be a hashmap, and returning the keys to said hashmap.

This is a leaky abstraction if ever I've seen one. And it's further confusing because it's unclear what exactly that hashmap is. The Persistence class takes a single argument which is the key that points to the map in question, but looking back at the constructor for the Persistence class and expects either a `context` argument to the class (whatever that means) or an already created Persistence instance which could also be anything.